src/builtins.rs

```diff
@@ -0,0 +1,686 @@
+            .replace("\\)", "%%R_PAREN%%")
+            .replace("|", "\\|")
+            .replace("(", "\\(")
+            .replace(")", "\\)")
+            .replace("%%PIPE%%", "|")
+            .replace("%%L_PAREN%%", "(")
+            .replace("%%R_PAREN%%", ")");
+        let regex = posix_regex::compile::PosixRegexBuilder::new(regex_str.as_bytes())
+            .with_default_classes()
+            .compile()
+            .expect("error compiling regex");
+        let param = match param.eval()?.borrow() {
+            NixValue::Str(x) => x.clone(),
+            x => return Err(EvalError::Unimplemented(format!("cannot cast {:?}", x))),
+        };
+        let matches = regex.matches(param.as_bytes(), None);
+
+        let mut out = vec![];
+
+        let mut cursor = 0;
+        for found in matches {
+            let (from, to) = found[0].ok_or(EvalError::Parsing)?;
+            out.push(Gc::new(Tree::from_concrete(NixValue::Str(
+                param[cursor..from].to_string(),
+            ))));
+            let mut here = vec![];
+            for group in found.iter().skip(1) {
+                here.push(match group {
+                    Some((x, y)) => {
+                        Gc::new(Tree::from_concrete(NixValue::Str(param[*x..*y].to_string())))
+                    }
+                    None => Gc::new(Tree::from_concrete(NixValue::Null)),
+                });
+            }
+            out.push(Gc::new(Tree::from_concrete(NixValue::List(here))));
+            cursor = to;
+        }
+        out.push(Gc::new(Tree::from_concrete(NixValue::Str(
+            param[cursor..].to_string(),
+        ))));
+        Ok(Gc::new(NixValue::List(out)))
+    }
+    "concatStringsSep" ; ConcatStringsSep => |param: Gc<Tree>| {
+        Ok(Gc::new(NixValue::Lambda(NixLambda::Builtin(NixBuiltin::ConcatStringsSep1(param.eval()?.as_str()?)))))
+    }
+    "concatStringsSep <?>" ; ConcatStringsSep1(_0: String) => |param: Gc<Tree>, sep: &String| {
+        let list = param.eval()?.as_list()?;
+        let mut out = vec![];
+        for item in list {
+            out.push(match item.eval()?.borrow() {
+                NixValue::Str(ref x) => x.clone(),
+                x => return Err(EvalError::Unimplemented(format!("cannot cast {:?}", x))),
+            });
+        }
+        Ok(Gc::new(NixValue::Str(out.join(sep))))
+    }
+    "unsafeGetAttrPos" ; UnsafeGetAttrPos => |param: Gc<Tree>| {
+        Ok(Gc::new(NixValue::Lambda(NixLambda::Builtin(NixBuiltin::UnsafeGetAttrPos1(match param.eval()?.borrow() {
+            NixValue::Str(x) => x.clone(),
+            x => return Err(EvalError::Unimplemented(format!("cannot cast {:?}", x))),
+        })))))
+    }
+    // FIXME: missing functionality
+    "unsafeGetAttrPos <?>" ; UnsafeGetAttrPos1(_0: String) => |_param: Gc<Tree>, _attr: &String| {
+        let mut map = HashMap::new();
+        map.insert("column".to_string(), Gc::new(Tree::from_concrete(
+            NixValue::Integer(1)
```

Comment:
This seems a little bit unexpected. Why not adding an `unimplemented`-error here? Or is this used too often to do this?

Reply:
Ah, iirc `unsafeGetAttrPor` is required for evaluating nixpkgs, so I needed to "implement" it to see if an evaluating lsp is even possible. Like many things, I will make sure to provide an actual implementation before putting it in any non-draft PRs to rnix-lsp.